### PR TITLE
Feature/File upload 2.0

### LINKED ISSUE #152: File upload 2.0
* You should be able to send a message before the upload is completed
* The upload should show the upload progress instead of the loading icon

### LINKED ISSUE #194: Denied permissions hides attachment button permanently
And granted permission from app settings doesn't make attachment button visible again:
Mon_Oct_21_01:22:56_CEST_2019

### LINKED ISSUE #247: Sent message progress is infinite.
Sometimes is hard to reproduce.

### PR DESCRIPTION
**Pull request by adrian09h**
# Submit a pull request(fixed #152, fixed #247  , fixed #194  )

- upload progress.
- send/edit message(locally) while uploading a file.
- add `editMessage` function to `ChannelViewModel`
- Message.setSyncStatus(Sync.SYNCED) on `ChatChannelEventHandler` and `ChatEventHandler`
```java
final void dispatchEvent(Client client, Event event) {   
        switch (event.getType()) {
            ...            
            case MESSAGE_NEW:
                event.getMessage().setSyncStatus(Sync.SYNCED);
                dispatchChannelEvent(client, event, this::onMessageNew);
                break;
            case MESSAGE_UPDATED:
                event.getMessage().setSyncStatus(Sync.SYNCED);
                dispatchChannelEvent(client, event, this::onMessageUpdated);
                break;
            case MESSAGE_DELETED:
                event.getMessage().setText(Constant.MESSAGE_DELETED);
                dispatchChannelEvent(client, event, this::onMessageDeleted);
                break;
           ...
        }
}
```

### CODE REVIEW

**Comment by tschellenbach** [comment]
icon for uploads somehow doesn't show for me...

image

File: library/src/main/java/com/getstream/sdk/chat/enums/MessageStatus.java [hunk 1]
```diff
@@ -3,7 +3,8 @@
 public enum MessageStatus {
     RECEIVED("received"),
     FAILED("failed"),
-    SENDING("sending");
+    SENDING("sending"),
+    PENDING("pending");
```
Comment (tbarbugli) [inline library/src/main/java/com/getstream/sdk/chat/enums/MessageStatus.java]: @tschellenbach does this make sense for you now that have syncStatus as well?

**Comment by adrian09h** [comment]
> icon for uploads somehow doesn't show for me...

Done

File: CHANGELOG.md [hunk 2]
```diff
@@ -15,6 +15,7 @@
 - Improved thread display
 - Expose ban information for current user (`User#getBanned`)
 - Bugfix on attachment size
+- Add `editMessage` functions to `ChannelViewModel` (Adrian)
```
Comment (tbarbugli) [inline CHANGELOG.md]: can you yank this out? makes it harder to merge

File: library/src/main/java/com/getstream/sdk/chat/adapter/MediaAttachmentSelectedAdapter.java [hunk 3]
```diff
@@ -75,6 +75,13 @@ public void bind(Attachment attachment, final OnItemClickListener listener) {
                     .solidColor(Color.BLACK)
                     .cornerRadii(cornerRadius, cornerRadius, cornerRadius, cornerRadius)
                     .build());
+
+            binding.ivMask.setShape(context, new DrawableBuilder()
```
Comment (tbarbugli) [inline library/src/main/java/com/getstream/sdk/chat/adapter/MediaAttachmentSelectedAdapter.java]: When I try to add a second file the dropdown is hidden

image

File: library/src/main/java/com/getstream/sdk/chat/adapter/MediaAttachmentSelectedAdapter.java [hunk 4]
```diff
@@ -75,6 +75,13 @@ public void bind(Attachment attachment, final OnItemClickListener listener) {
...
+                    .rectangle()
```
Comment (tbarbugli) [inline library/src/main/java/com/getstream/sdk/chat/adapter/MediaAttachmentSelectedAdapter.java]: File picker break if you pick one file and then try to pick another one. See screenshot for final state (I picked 1 image and then tapped on the same again)

image

File: library/src/main/java/com/getstream/sdk/chat/adapter/MediaAttachmentSelectedAdapter.java (hunk 3 above)
Comment (tbarbugli) [reply library/src/main/java/com/getstream/sdk/chat/adapter/MediaAttachmentSelectedAdapter.java]: Somehow I manage to do this on the simulator

image

When this happens the image is selected forever

Comment (tbarbugli) [inline library/src/main/java/com/getstream/sdk/chat/adapter/MediaAttachmentSelectedAdapter.java]: When uploading a file fails (ie too large) the file stays selected and clicking to unselect actually puts it back in the message (which will fail)

image

Comment (tbarbugli) [inline library/src/main/java/com/getstream/sdk/chat/adapter/MediaAttachmentSelectedAdapter.java]: Length preview is still wrong (video is a couple minutes)

image

Comment (adrian09h) [reply library/src/main/java/com/getstream/sdk/chat/adapter/MediaAttachmentSelectedAdapter.java]: This issue was fixed on this PR.
Please try this branch.

File: library/src/main/java/com/getstream/sdk/chat/enums/MessageStatus.java (hunk 1 above)
Comment (adrian09h) [reply library/src/main/java/com/getstream/sdk/chat/enums/MessageStatus.java]: I have removed `MessageStatus` and `MessageStatusConverter`, We can handle with `Sync` for `MessageStatus` behavior.

File: CHANGELOG.md (hunk 2 above)
Comment (adrian09h) [reply CHANGELOG.md]: Resolved conflicts the line.

File: library/src/main/java/com/getstream/sdk/chat/adapter/MediaAttachmentSelectedAdapter.java (hunk 3 above)
Comment (adrian09h) [reply library/src/main/java/com/getstream/sdk/chat/adapter/MediaAttachmentSelectedAdapter.java]: Fixed it.

Comment (adrian09h) [reply library/src/main/java/com/getstream/sdk/chat/adapter/MediaAttachmentSelectedAdapter.java]: I can't reproduce the issue in my side.
Can you please send me the problem video?

File: library/src/main/java/com/getstream/sdk/chat/adapter/MediaAttachmentSelectedAdapter.java (hunk 4 above)
Comment (adrian09h) [reply library/src/main/java/com/getstream/sdk/chat/adapter/MediaAttachmentSelectedAdapter.java]: Fixed it

File: library/src/main/java/com/getstream/sdk/chat/rest/controller/RetrofitClient.java [hunk 5]
```diff
@@ -58,9 +58,6 @@ public static Retrofit getAuthorizedClient(CachedTokenProvider tokenProvider, Ap
 
     public static Retrofit getAuthorizedCDNClient(CachedTokenProvider tokenProvider, ApiClientOptions options) {
 
-        HttpLoggingInterceptor loggingInterceptor = new HttpLoggingInterceptor();
```
[the remark this answers is not in the document]
Comment (elevenetc) [reply library/src/main/java/com/getstream/sdk/chat/rest/controller/RetrofitClient.java]: Wonder why it was removed? It's useful to see http logs.

File: library/src/main/java/com/getstream/sdk/chat/view/MessageInputStyle.java [hunk 6]
```diff
@@ -134,14 +134,15 @@ private Drawable getSelector(@ColorInt int normalColor, @ColorInt int pressedCol
...
+//        return prefs.getBoolean(showAttachmentButtonKey, showAttachmentButton);
+        return showAttachmentButton;
     }
 
-    public void setShowAttachmentButton(boolean showAttachmentButton) {
-        this.showAttachmentButton = showAttachmentButton;
-        prefs.edit().putBoolean(showAttachmentButtonKey, showAttachmentButton).apply();
-        prefs.edit().putBoolean(permissionSetKey, true).apply();
-    }
+//    public void setShowAttachmentButton(boolean showAttachmentButton) {
```
Comment (elevenetc) [inline library/src/main/java/com/getstream/sdk/chat/view/MessageInputStyle.java]: Should be removed if it's not needed

File: library/src/main/java/com/getstream/sdk/chat/view/MessageInputView.java [hunk 7]
```diff
@@ -374,7 +399,7 @@ public void permissionResult(int requestCode, @NonNull String[] permissions,
                 }
             }
 
-            style.setShowAttachmentButton(granted);
+//            style.setShowAttachmentButton(granted);
```
Comment (elevenetc) [inline library/src/main/java/com/getstream/sdk/chat/view/MessageInputView.java]: Should be removed if it's not needed.

File: library/src/main/java/com/getstream/sdk/chat/rest/controller/RetrofitClient.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (adrian09h) [reply library/src/main/java/com/getstream/sdk/chat/rest/controller/RetrofitClient.java]: @elevenetc It is not suitable for file uploading with checking `onProgress`, I think

https://stackoverflow.com/questions/40691610/retrofit-2-requestbody-writeto-method-called-twice

[the remark this answers is not in the document]
Comment (elevenetc) [reply library/src/main/java/com/getstream/sdk/chat/rest/controller/RetrofitClient.java]: Hm, that's bad.

**Pull request closed without merging**